src/Tools/ExternalAccess/Razor/Features/RazorAnalyzerAssemblyResolver.cs

```diff
@@ -49,63 +49,84 @@ internal sealed class RazorAnalyzerAssemblyResolver() : IAnalyzerAssemblyResolve
                 return null;
             }
 
-            var assembly = compilerLoadContext.Assemblies.FirstOrDefault(a => a.GetName().Name == assemblyName.Name);
-            if (assembly is not null)
-            {
-                return assembly;
-            }
+            // load the complete closure of razor assemblies if we're asked to load any of them. Subsequent requests for the others will just return the ones loaded here
+            LoadAssemblyByFileName(compilerLoadContext, RazorCompilerAssemblyName, directory);
+            LoadAssemblyByFileName(compilerLoadContext, RazorUtilsAssemblyName, directory);
+            LoadAssemblyByFileName(compilerLoadContext, ObjectPoolAssemblyName, directory);
 
-            var assemblyFileName = $"{assemblyName.Name}.dll";
+            // return the actual assembly that we were asked to load.
+            return LoadAssembly(compilerLoadContext, assemblyName, directory);
 
-            // Depending on who wins the race to load these assemblies, the base directory will either be the tooling root (if Roslyn wins)
-            // or the ServiceHubCore subfolder (razor). In the root directory these are netstandard2.0 targeted, in ServiceHubCore they are 
-            // .net targeted. We need to always pick the same set of assemblies regardless of who causes us to load. Because this code only
-            // runs in a .net based host, it's safe to always choose the .net targeted ServiceHubCore versions.
-            if (!Path.GetFileName(directory.AsSpan().TrimEnd(Path.DirectorySeparatorChar)).Equals(ServiceHubCoreFolderName, StringComparison.OrdinalIgnoreCase))
+            static Assembly? LoadAssemblyByFileName(AssemblyLoadContext compilerLoadContext, string fileName, string directory)
             {
-                var serviceHubCoreDirectory = Path.Combine(directory, ServiceHubCoreFolderName);
-
-                // The logic above only applies to VS. In VS Code there is no service hub, so appending the folder would be silly.
-                if (Directory.Exists(serviceHubCoreDirectory))
+                var onDiskName = Path.Combine(directory, $"{fileName}.dll");
```

Comment:
This is kind of odd that we find the assembly on disk, read it to get its assemblyName, then load it by assemblyName, which in turn attempts to find it on the disk, but ensures we go through the correct redirection etc later on.

Reply:
Can you add that as a comment?